### PR TITLE
Support LingV2_5 model

### PR DESCRIPTION
**Pull request by ant-yy**
## Motivation
This is used for support Bailing models, Ling-2.5 and Ring-2.5

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/configs/bailing_hybrid.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            if type_value == HybridLayerType.full_attention.value
+        ]
+
+    @property
+    def mamba2_cache_params(self) -> Mamba2CacheParams:
+        from sglang.srt.layers.dp_attention import get_attention_tp_size
+
+        shape = Mamba2StateShape.create(
+            tp_world_size=get_attention_tp_size(),
+            intermediate_size=0,  # 使用纯linear_kv, 没有卷积
```
Comment (ispobock) [inline python/sglang/srt/configs/bailing_hybrid.py]: Could you use english comments?

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 2]
```diff
@@ -1125,6 +1135,359 @@ def forward_extend(self, *args, **kwargs):
...
+        layer_id = layer.layer_id if layer else kwargs["layer_id"]
+
+        # Use precomputed metadata across all layers
+        metadata = self.forward_metadata
+
+        if self.kv_cache_dtype_str != "auto":
+            q = q.to(self.kv_cache_dtype)
+
+        # Do linear attention
+        query_start_loc = self.forward_metadata.query_start_loc  # 确认下shape
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: update comments

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 3]
```diff
@@ -1125,6 +1135,359 @@ def forward_extend(self, *args, **kwargs):
...
+                    seq_len = forward_batch.extend_seq_lens[_prefill_idx]
+                    _end = _start + seq_len
+                else:
+                    _end = q.shape[0]
+
+            slot_id = state_indices_tensor[_prefill_idx]
+            qs = q[_start:_end].transpose(0, 1).contiguous()
+            ks = k[_start:_end].transpose(0, 1).contiguous()
+            vs = v[_start:_end].transpose(0, 1).contiguous()
+            slice_layer_cache = kv_cache[slot_id, ...]  # 获取现在的kvcache
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: update comments

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 4]
```diff
@@ -1125,6 +1135,359 @@ def forward_extend(self, *args, **kwargs):
...
+        q_rope = kwargs["q_rope"] if "q_rope" in kwargs else None
+        k_rope = kwargs["k_rope"] if "k_rope" in kwargs else None
+        layer_id = layer.layer_id if layer else kwargs["layer_id"]
+
+        metadata = self.forward_metadata
+
+        if self.kv_cache_dtype_str != "auto" and layer.k_scale is not None:
+            q = q.to(self.kv_cache_dtype)
+
+        query_start_loc = self.forward_metadata.query_start_loc  # 确认下shape
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: update comments

File: python/sglang/srt/configs/bailing_hybrid.py (hunk 1 above)
Comment (ispobock) [inline python/sglang/srt/configs/bailing_hybrid.py]: update comments

File: python/sglang/srt/layers/attention/linear/rmsnorm.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        weight.stride(0),
+        rstd.stride(0),
+        eps,
+        BLOCK_N,
+    )
+    # residual_out is None if residual is None and residual_dtype == input_dtype
+    return y, rstd
+
+
+class RMSNormTriton(torch.autograd.Function):
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/linear/rmsnorm.py]: Can we use existing module? https://github.com/sgl-project/sglang/blob/feaa9e7e00e624eb25e375e81fd5d47c78080874/python/sglang/srt/layers/layernorm.py#L89

File: python/sglang/srt/layers/attention/linear/linear_rotary_embedding.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
+# Adapted from https://raw.githubusercontent.com/vllm-project/vllm/refs/tags/v0.6.6.post1/vllm/model_executor/layers/rotary_embedding.py
+
+"""Rotary Positional Embeddings for support fp32."""
+from typing import Any, Dict, Optional, Tuple
+
+import torch
+from vllm._custom_ops import rotary_embedding
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/linear/linear_rotary_embedding.py]: We cannot import vllm custom ops.
Could you use sgl-kernel instead?

File: python/sglang/srt/layers/attention/linear/linear_rotary_embedding.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+from sglang.srt.layers.rotary_embedding import (
+    RotaryEmbedding,
+    _yarn_find_correction_range,
+    _yarn_get_mscale,
+    _yarn_linear_ramp_mask,
+)
+from sglang.srt.server_args import get_global_server_args
+
+
+class Fp32RotaryEmbedding(RotaryEmbedding):
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/linear/linear_rotary_embedding.py]: What's the difference between this Rope module and https://github.com/sgl-project/sglang/blob/feaa9e7e00e624eb25e375e81fd5d47c78080874/python/sglang/srt/layers/rotary_embedding.py#L99
Can we reuse or extend the existing one?

File: python/sglang/srt/layers/attention/linear/linear_metadata.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
+from dataclasses import dataclass
+
+import torch
+
+from sglang.srt.layers.attention.mamba.mamba2_metadata import ForwardMetadata
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+
+
+@dataclass(kw_only=True)
+class BailingLinearMetadata(ForwardMetadata):
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/linear/linear_metadata.py:10]: I'm not sure if it's good to put model specific metadata here. @yizhang2077 Could you double check?

File: python/sglang/srt/managers/scheduler.py [hunk 9]
```diff
@@ -616,6 +616,7 @@ def init_cache_with_memory_pool(self):
         self.is_hybrid_ssm = (
             self.tp_worker.model_runner.hybrid_gdn_config is not None
             or self.tp_worker.model_runner.mamba2_config is not None
+            or self.tp_worker.model_runner.hybrid_gdn_config is not None
```
Comment (yizhang2077) [inline python/sglang/srt/managers/scheduler.py]: duplicate logic? maybe it is `hybrid_lightning_config`?

File: python/sglang/srt/layers/attention/linear/linear_rotary_embedding.py (hunk 6 above)
Comment (ant-yy) [reply python/sglang/srt/layers/attention/linear/linear_rotary_embedding.py]: We use sgl-kernel instead in the latest commit

File: python/sglang/srt/layers/attention/linear/linear_metadata.py (hunk 8 above)
Comment (ant-yy) [reply python/sglang/srt/layers/attention/linear/linear_metadata.py:10]: We followed the code structure of Mamba as a reference—Mamba also places its model-specific metadata (e.g., mamba2_metadata.py) within the mamba/ directory.

**Review by ispobock — APPROVED** [review]
